src/pybind/mgr/ansible_orchestrator/module.py

```diff
@@ -0,0 +1,364 @@
+        """
+
+        self.pb_execution.launch()
+
+    def update_result(self):
+        """Output of the read operation
+
+        The result of the playbook execution can be customized through the
+        function provided as 'process_output' attribute
+
+        @return string: Result of the operation formatted if it is possible
+        """
+
+        processed_result = []
+
+        if self._is_complete:
+            raw_result = self.pb_execution.get_result(self.event_filter)
+
+            if self.process_output:
+                processed_result = self.process_output(
+                                            raw_result,
+                                            self.ar_client,
+                                            self.pb_execution.play_uuid)
+            else:
+                processed_result = raw_result
+
+        self._result = processed_result
+
+
+class AnsibleChangeOperation(orchestrator.WriteCompletion):
+    """Operations that changes the "cluster" state
+
+    Modifications/Changes (writes) are a two-phase thing, firstly execute
+    the playbook that is going to change elements in the Ceph Cluster.
+    When the playbook finishes execution (independently of the result),
+    the modification/change operation has finished.
+    """
+    def __init__(self):
+        super(AnsibleChangeOperation, self).__init__()
+
+        self.error = False
+    @property
+    def status(self):
+        """Return the status code of the operation
+        """
+        #TODO
+        return 0
+
+    @property
+    def is_persistent(self):
+        """
+        Has the operation updated the orchestrator's configuration
+        persistently?  Typically this would indicate that an update
+        had been written to a manifest, but that the update
+        had not necessarily been pushed out to the cluster.
+
+        In the case of Ansible is always False.
+        because a initiated playbook execution will need always to be
+        relaunched if it fails.
+        """
+
+        return False
+
+    @property
+    def is_effective(self):
+        """Has the operation taken effect on the cluster?
+        For example, if we were adding a service, has it come up and appeared
+        in Ceph's cluster maps?
+
+        In the case of Ansible, this will be True if the playbooks has been
+        executed succesfully.
+
+        @return Boolean: if the playbook has been executed succesfully
+        """
+
+        return self.status == ExecutionStatusCode.SUCCESS
+
+    @property
+    def is_errored(self):
+        return self.error
+
+    @property
+    def is_complete(self):
+        return self.is_errored or (self.is_persistent and self.is_effective)
+
+
+class Module(MgrModule, orchestrator.Orchestrator):
+    """An Orchestrator that an external Ansible runner service to perform
+    operations
+    """
+
+    OPTIONS = [
+        {'name': 'server_addr'},
+        {'name': 'server_port'},
+        {'name': 'username'},
```

Comment:
Does ansible-runner-service actually have/need multi-tenancy (multiple user accounts)?  It feels like an unnecessary complexity when the service is just for a single cluster.

Reply:
This is how Ansible runner service works. Even in a single cluster multiple users with different privileges use to exist.

Reply:
Some of the following may need input from @pcuzner on the intended security model:

I don't think that ARS has users with different privileges.  From reading https://github.com/pcuzner/ansible-runner-service/blob/master/runner_service/controllers/login.py#L46 it seems like they just have a password stored in plain text in their config file, and once you're logged in it doesn't matter what user you are.

Even if ARS did expand its user account concept beyond a dict in the config file, all the ansible playbooks are being run as root out on the cluster nodes, so would it be any meaningful security isolation?

BTW, I also notice that ansible_runner_service has a default crypto secret of "secret" and nothing in the installation instructions about how to set it to something unique per-installation, so hopefully there is a plan for resolving that.

I also don't see any mechanism for revoking JWT tokens, so it seems like even if a user account was removed from the ARS configuration file, login sessions would continue until expiry (default 24 hours).

The username/password handling seems quite superficial, so I'm left wondering why we don't just use the client TLS certificates for authentication.  The user/pass stuff seems like it's liable to give a false sense of security -- if the certs are handled properly, they should be enough security.

Reply:
There aren't users with different privileges in Ansible Runner Service... but probably users will want a certain level of security. (what i mean is that not all the users that can access/use the servers has the same possibilities of doing things) 

Ansible Runner Service is quite new ( like me) so we need a little time in order to be completely functional.  :-) 
By the moment what we have is the "user login" and the use of tokens... and as you said ... we have several ways to improve this point.